pkg/api/rest/update.go

```diff
@@ -103,3 +104,35 @@ func BeforeUpdate(strategy RESTUpdateStrategy, ctx api.Context, obj, old runtime
 
 	return nil
 }
+
+type defaultUpdatedObjectInfo struct {
+	obj runtime.Object
+}
+
+func DefaultUpdatedObjectInfo(obj runtime.Object) UpdatedObjectInfo {
+	return &defaultUpdatedObjectInfo{obj}
+}
+
+func (i *defaultUpdatedObjectInfo) Preconditions() *api.Preconditions {
+	preconditions := &api.Preconditions{}
+
+	// Attempt to get the UID out of the object
+	accessor, err := meta.Accessor(i.obj)
+	if err != nil {
+		return preconditions
+	}
+
+	// If non-empty, set as a precondition
+	uid := accessor.GetUID()
+	if len(uid) > 0 {
+		preconditions.UID = &uid
+	}
+
+	return preconditions
+}
+
+func (i *defaultUpdatedObjectInfo) UpdatedObject(ctx api.Context, oldObj runtime.Object) (runtime.Object, error) {
+	// Don't return the original, since BeforeUpdate can mutate this, doing things like clearing ResourceVersion.
+	// If we're re-called, we need to be able to return the pristine version.
+	return api.Scheme.Copy(i.obj)
```

Comment:
This resolves a TODO inside the generic store that noted that all sorts of things in the updated object got mutated in BeforeCreate. It also exposed several tests that only passed because of side-effect modifications on the object passed to Update that made later DeepEqual comparisons pass.